wal/src/message_queue_impl/namespace.rs

```diff
@@ -0,0 +1,714 @@
+
+        let region = self
+            .get_or_open_region(location.shard_id as RegionId)
+            .await
+            .context(MarkDeleteTo {
+                namespace: self.namespace.clone(),
+                location,
+                sequence_num,
+            })?;
+
+        region
+            .mark_delete_to(location.table_id, sequence_num)
+            .await
+            .context(MarkDeleteTo {
+                namespace: self.namespace.clone(),
+                location,
+                sequence_num,
+            })
+    }
+
+    async fn clean_logs(&self) -> Result<()> {
+        let regions = { self.regions.read().await.clone() };
+
+        for region in regions.values() {
+            region.clean_logs().await.context(CleanLogs {
+                namespace: self.namespace.clone(),
+            })?;
+        }
+
+        Ok(())
+    }
+}
+
+type RegionRef<M> = Arc<Region<M>>;
+
+/// Iterator of scanning the whole region
+#[derive(Debug)]
+pub struct ScanRegionIterator<C: ConsumeIterator> {
+    /// Namespace name
+    namespace: String,
+
+    /// Log iterator
+    ///
+    /// It can be `None`, while there isn't any log of the table in the region
+    /// actually.
+    iter: Option<MessageQueueLogIterator<C>>,
+
+    /// Request triggering this scanning
+    request: ScanRequest,
+
+    /// Terminated flag
+    is_terminated: bool,
+
+    /// See the same problem in https://github.com/CeresDB/ceresdb/issues/120
+    previous_value: Vec<u8>,
+}
+
+impl<C: ConsumeIterator> ScanRegionIterator<C> {
+    fn new_empty() -> Self {
+        Self {
+            namespace: String::default(),
+            iter: None,
+            request: ScanRequest::default(),
+            is_terminated: true,
+            previous_value: Vec::new(),
+        }
+    }
+
+    fn new(namespace: String, iter: MessageQueueLogIterator<C>, request: ScanRequest) -> Self {
+        Self {
+            namespace,
+            iter: Some(iter),
+            request,
+            previous_value: Vec::new(),
+            is_terminated: false,
+        }
+    }
+
+    pub async fn next_log_entry(&mut self) -> Result<Option<LogEntry<&'_ [u8]>>> {
+        // If inner iter is `Some`, poll it to get required log entry.
+        if let Some(iter) = self.iter.as_mut() {
+            if self.is_terminated {
+                return Ok(None);
+            }
+
+            iter.next_log_entry()
+                .await
+                .context(ScanWithCause {
+                    namespace: self.namespace.clone(),
+                    request: self.request.clone(),
+                    msg: "failed while polling log",
+                })
+                .map(|log_entry| {
+                    if log_entry.is_none() {
+                        self.is_terminated = true;
+                    }
+
+                    log_entry.map(|log_entry| {
+                        self.previous_value = log_entry.payload.to_owned();
```

Comment:
it seems no need to call `to_owned()`?

Reply:
It needs, the borrow check isn't smart enough to know the &[u8] is actually owed by current iterator ... So in last pr about this, I plan to return `Vec<u8>` instead of `&[u8]` from the inner iterator.

Actually, return &[u8] is just a effective optimization in rocksdb implementation. I found it make trouble in other implements temporarily.